lib/signers/v4_credentials.js

```diff
@@ -0,0 +1,95 @@
+var AWS = require('../core');
+
+/**
+ * @api private
+ */
+var cachedSecret = {};
+
+/**
+ * @api private
+ */
+var cacheQueue = [];
+
+/**
+ * @api private
+ */
+var maxCacheEntries = 50;
+
+/**
+ * @api private
+ */
+var v4Identifier = 'aws4_request';
+
+AWS.Signers.v4_credentials = {
+  /**
+   * @api private
+   *
+   * @param date [String]
+   * @param region [String]
+   * @param serviceName [String]
+   * @returns {string}
+   */
+  createScope: function createScope(date, region, serviceName) {
+    return [
+      date.substr(0, 8),
+      region,
+      serviceName,
+      v4Identifier
+    ].join('/');
+  },
+
+  /**
+   * @api private
+   *
+   * @param credentials [Credentials]
+   * @param date [String]
+   * @param region [String]
+   * @param service [String]
+   * @param shouldCache [Boolean]
+   * @returns {String}
+   */
+  getSigningKey: function getSigningKey(
+    credentials,
+    date,
+    region,
+    service,
+    shouldCache
+  ) {
+    var cacheKey = [credentials.accessKeyId, date, region, service].join('_');
```

Comment:
We had 2 issues reported because this cache key wasn't unique enough:
https://github.com/aws/aws-sdk-js/pull/1054

You may need to pass the serviceClientId as well.